core/opendaq/opendaq/src/instance_impl.cpp

```diff
@@ -26,6 +26,62 @@ InstanceImpl::InstanceImpl(ContextPtr context, const StringPtr& localId)
     loggerComponent = this->context.getLogger().addComponent("Instance");
 }
 
+static ContextPtr ContextFromInstanceBuilder(IInstanceBuilder* instanceBuilder)
+{
+    const auto builderPtr = InstanceBuilderPtr::Borrow(instanceBuilder);
+
+    auto logger = builderPtr.getLogger();
+    auto scheduler = builderPtr.getScheduler();
+    auto moduleManager = builderPtr.getModuleManager();
+    auto typeManager = TypeManager();
+
+    // Configure logger
+    if (!logger.assigned()) 
+    {
+        auto sinks = builderPtr.getLoggerSinks();
+        if (sinks.empty())
+            logger = Logger();
+        else
+            logger = LoggerWithSinks(sinks);
+        logger.setLevel(builderPtr.getGlobalLogLevel());
+    }
+
+    for (const auto& [component, logLevel] : builderPtr.getComponentsLogLevel())
+    {
+        auto createdComponent = logger.getOrAddComponent(component);
+        createdComponent.setLevel(LogLevel(logLevel));
+    }
+
+    // Configure scheduler
+    if (!scheduler.assigned())
+        scheduler = Scheduler(logger, builderPtr.getSchedulerWorkerNum());
+
+    // Configure moduleManager
+    if (!moduleManager.assigned())
+        moduleManager = ModuleManager(builderPtr.getModulePath());
+
+    return Context(scheduler, logger, typeManager, moduleManager);
+}
+
+InstanceImpl::InstanceImpl(IInstanceBuilder* instanceBuilder)
+    : context(ContextFromInstanceBuilder(instanceBuilder))
+    , moduleManager(this->context.assigned() ? this->context.asPtr<IContextInternal>().moveModuleManager() : nullptr)
+    , rootDeviceSet(false)
+{
+    const auto builderPtr = InstanceBuilderPtr::Borrow(instanceBuilder);
+    
+    auto localId = builderPtr.getDefaultRootDeviceLocalId();
+    auto instanceId = defineLocalId(localId.assigned() ? localId.toStdString() : std::string());
+    defaultRootDevice = Client(this->context, instanceId, builderPtr.getDefaultRootDeviceInfo());
+    rootDevice = defaultRootDevice;
+
+    auto connectionString = builderPtr.getRootDevice();
+    if (connectionString.assigned())
+        rootDevice = detail::createDevice(connectionString, nullptr, nullptr, moduleManager, nullptr);    
```

Comment:
Nitpicking, but the `defaultRootDevice` should probably not be assigned to `rootDevice` at all if the `connectionString` is assigned. I'd also add another check for the `connectionString` being empty and ignore it in that case.